File: leetcodes/128. Longest Consecutive Sequence/longestConsecutiveSeq.java
class Solution {
  public int longestConsecutive(int[] nums) {
    /**
     * 1) brute force: keeps checking if num + 1 in nums
     *      time O(n^3), space O(1)
     *
     * 2) sorting -> O(nlogn) space O(n) or O(1)
     *
     * 3) hashSet and check if in seq -> timeO(n) space O(n)
     *
     */
    Set<Integer> set = new HashSet<>();
    for (int n : nums) {
      set.add(n);
    }
    int best = 0;
    for (int n : set) {
      if (!set.contains(n-1)){
        int m = n + 1;
        while(set.contains(m)) {
          m++;
        }
        best = Math.max(best, m - n);
      }
    }
    return best;
  }
}

File: leetcodes/128. Longest Consecutive Sequence/longestConsecutiveSeq.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
      '''
        sort time nlog(n) ->iterate over the sorted array to check if consecutive and then increase the count
        space O(1) or O(n) to store new copy
      '''
      if not nums: return 0
      nums.sort()
      count = 1
      maxC = 1
      prev = nums[0]
      for i in range(1,len(nums)):
        if nums[i] == prev + 1:
          count += 1
        elif nums[i] == nums[i-1]:
          prev = nums[i]
        else:
          count = 1
        prev = nums[i]
        maxC = max(maxC, count)
      return maxC

      #hashset timeO(n) spaceO(n) -> but run time is slower than sorting
